### DISCUSSION TITLE
How to handle the door/window connector

### DISCUSSION THREAD

**Discussion by Fr3d3rick83** [windkh/node-red-contrib-shelly#105] [Q&A] [answered]
Hi,
my NR runs in a Conatiner and I have this flow:
```
[
    {
        "id": "bb2a887329abef75",
        "type": "shelly-gen1",
        "z": "8b59e77cd35706bb",
        "hostname": "192.168.178.93",
        "description": "Haustür",
        "mode": "callback",
        "server": "de700d896c3165e8",
        "outputmode": "status",
        "uploadretryinterval": 5000,
        "pollinginterval": 5000,
        "pollstatus": false,
        "getstatusoncommand": true,
        "devicetype": "Sensor",
        "outputs": 1,
        "x": 300,
        "y": 80,
        "wires": [
            [
                "5c171bcb0dffb583"
            ]
        ]
    },
    {
        "id": "5c171bcb0dffb583",
        "type": "debug",
        "z": "8b59e77cd35706bb",
        "name": "debug 1",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "false",
        "statusVal": "",
        "statusType": "auto",
        "x": 480,
        "y": 80,
        "wires": []
    },
    {
        "id": "fabe9f0716b48eb1",
        "type": "inject",
        "z": "8b59e77cd35706bb",
        "name": "",
        "props": [
            {
                "p": "payload"
            },
            {
                "p": "topic",
                "vt": "str"
            }
        ],
        "repeat": "",
        "crontab": "",
        "once": false,
        "onceDelay": 0.1,
        "topic": "",
        "payload": "",
        "payloadType": "date",
        "x": 120,
        "y": 80,
        "wires": [
            [
                "bb2a887329abef75"
            ]
        ]
    },
    {
        "id": "de700d896c3165e8",
        "type": "shelly-gen1-server",
        "port": "6335",
        "hostname": ""
    }
]
```
My NR adress is: http://192.168.178.57:6335. So how can I store and update the current status of the connector? It says always "timeout of 5000ms exceeded".

**Comment by Fr3d3rick83** [comment]
Any idea, how to send a signal to Node Red?

**Reply by windkh** [reply]
Maybe Haustür sleeps... wake it up and redeploy. The rest looks fine.

Port 6335 must be configured in docker so that the node can listen on this port.

**Reply by Fr3d3rick83** [reply]
But the sensor is almost all the time asleep, to save power. It only awakes, wenn you open or close something? So if its not working, maybe over MQTT.

**Reply by windkh** [reply]
The node retries to set the webhooks or callbacks until it succeeds. I have this setup here running without any issues.
Have you tested the flow outside of docker yet? If it works there then it is a docker configuration issue.

**Comment by Fr3d3rick83** [comment]
How does the port needs to be configured in Docker? The Sensor is working perfect within the Shelly App.

**Reply by windkh** [reply]
The port 1880 of node red and the one that you configured as callback must be exposed in your dockerfile. 
If you switch to polling. Does it work then?

**Reply by Fr3d3rick83** [reply]
Ok, how do I expose this in the dockerfile?
Callback is working but only, when I open the door and than push the trigger in NodeRed, than it says connected. but when I trigger it again, it says timeout.

**Comment by windkh** [comment]
@Fr3d3rick83
1. expose port in docker file: just search for the line where the port 1880 of node red is configured. You should do it the same way with the port you are using for your server-node. I guess in your case it is 6335.
2. Find out the IP address under that your docker instance is reachable from within your local LAN and put this IP address in the field hostname of your server node: 
`
{
        "id": "de700d896c3165e8",
        "type": "shelly-gen1-server",
        "port": "6335",
        "hostname": "<your IP address here>"
    }
`
3. check if the node configured your shelly the right way. Just open the device in your shelly app and have a look at the IO Actions:
![image](https://user-images.githubusercontent.com/10910767/205573612-08e16dd1-bad0-4261-a128-029c10ba780e.png)

The IP and PORT inside the URL in the middle should match the IP of your docker container.

4. if you want to check if the url triggers your docker image just copy it to a browser and reload:
`
http://192.168.178.200:10002/webhook?data=report_url?0?192.168.178.249
`
In this case 192.168.178.200:10002 is the docker image and 192.168.178.249 is the IP of your shelly.

If eyerything works then the flow should show something like:

![image](https://user-images.githubusercontent.com/10910767/205575066-998fe9f3-65d4-41bc-aaa6-0a23b6c05341.png)

**Reply by Fr3d3rick83** [reply]
Ok, I edit the node:
![Bildschirmfoto 2022-12-05 um 10 53 52](https://user-images.githubusercontent.com/97901424/205607971-be22f7fa-376a-426b-91e7-1f82d3b01a9e.png)

And changed the URL at the IO section, I dont know why but there was a wrong IP adress for the Docker
![Bildschirmfoto 2022-12-05 um 10 54 11](https://user-images.githubusercontent.com/97901424/205608112-cc43c1ad-a681-4f04-b74b-dbc6a4ff94b4.png)

When I try to open it in browser:
http://192.168.178.57:6335/webhook?data=report_url?0?192.168.178.93

It says:
Cannot GET /webhook

What function is written in the "Alarm" and "send markdown" node?

**Comment by windkh** [comment]
@Fr3d3rick83 configuring docker depends on what environment you are using. I for myself do all tests using docker on my QNAP NAS.
Here you can see 1880 as standard port and the others are just exposed the same way. In your docker file it should be similar to that.
Another detail is how you configured your docker network: I just use NAT but you can also expose the complete docker image as if it would be a separate node in your network. Hope this helps. 

![image](https://user-images.githubusercontent.com/10910767/205577076-9e69a6b3-31c2-43f1-b37f-874d1ba8429e.png)

**Reply by Fr3d3rick83** [reply]
I´m using Synology DSM Docker, it looks now like this:
![Bildschirmfoto 2022-12-05 um 10 27 31](https://user-images.githubusercontent.com/97901424/205602138-18b655b4-2049-4cb7-952e-38bbdcd41056.png)

**Reply by windkh** [reply]
> I´m using Synology DSM Docker, it looks now like this:
> 
> ![Bildschirmfoto 2022-12-05 um 10 27 31](https://user-images.githubusercontent.com/97901424/205602138-18b655b4-2049-4cb7-952e-38bbdcd41056.png)
> 
> 

Links ist der Fehler bei Porteinstellungen.
Dort bitte
Lokeler Port 1880 auf 1880 und in einer neuen Zeile drunter
Lokaler Port 6335 auf 6335 einstellen.

**Reply by Fr3d3rick83** [reply]
Also so?
![Bildschirmfoto 2022-12-05 um 13 17 33](https://user-images.githubusercontent.com/97901424/205635424-2c00c9eb-6f00-4706-bd69-b692e647bc2d.png)

**Reply by Fr3d3rick83** [reply]
So kann ich leider nur Norde Red nicht mehr aufrufen.
![Bildschirmfoto 2022-12-05 um 13 19 53](https://user-images.githubusercontent.com/97901424/205635840-7b0e5aa7-4e25-4c4d-a217-f4f8980402ef.png)

**Reply by windkh** [reply]
> Also so? ![Bildschirmfoto 2022-12-05 um 13 17 33](https://user-images.githubusercontent.com/97901424/205635424-2c00c9eb-6f00-4706-bd69-b692e647bc2d.png)

ja so

**Reply by windkh** [reply]
> So kann ich leider nur Norde Red nicht mehr aufrufen. ![Bildschirmfoto 2022-12-05 um 13 19 53](https://user-images.githubusercontent.com/97901424/205635840-7b0e5aa7-4e25-4c4d-a217-f4f8980402ef.png)

wieso? was geht nicht? container neu gestartet?

**Reply by Fr3d3rick83** [reply]
Wenn ich NodeRed aufrufe, kommt dieser Fehler von oben. Container ein paar mal neugestartet.
Hatte ihn damals so erstellt:
https://mariushosting.com/how-to-install-node-red-on-your-synology-nas/

Wenn ich die Ports wieder auf 6335:1880 setzte, kann ich nach einem Neustart wieder NodeRed aufrufen.

**Reply by windkh** [reply]
> > Also so? ![Bildschirmfoto 2022-12-05 um 13 17 33](https://user-images.githubusercontent.com/97901424/205635424-2c00c9eb-6f00-4706-bd69-b692e647bc2d.png)
> 
> ja so

So bitte einstellen und dann tut alles wenn du node-red einfach mit 1880 statt mit 6335 öffnest. 
http://synologynas:1880 

In der Anleitung von https://mariushosting.com/how-to-install-node-red-on-your-synology-nas/ wird halt einfach der port vom container auf einen andere umgebogen, das muss aber nicht sein sondern verwirrt eher.
Du willst aber der 6335 für die Kommunikation mit dem Shelly nutzen.

Also 
Lokaler Port --> Container Port
1880 --> 1880
6335 --> 6335

Dann node-red öffnen
http://synologynas:1880 
 
Neu deployen und checken ob es nun geht

**Reply by Fr3d3rick83** [reply]
Ja cool, dass klappt so, DANKE.
Passt der Rest so?
![Bildschirmfoto 2022-12-06 um 06 55 20](https://user-images.githubusercontent.com/97901424/205830162-32f57bc0-6ae2-4827-b1f2-80fc4f423e34.png)

Im Shelly nehme ich dann aber den anderen Port:
http://192.168.178.57:6335/webhook?data=report_url?0?192.168.178.93

Wie erstelle ich jetzt den Flow von oben?

**Reply by windkh** [reply]
@Fr3d3rick83 
Für die flows gibts die Beispiele im examples folder. 

Hier ein kleines Beispiel: bitte die chat id entsprechend abändern

[{"id":"13474f6d.cfac91","type":"telegram sender","z":"5882ea6c.1d8c3c","name":"","bot":"59e305c4.e5a97c","haserroroutput":false,"outputs":1,"x":830,"y":80,"wires":[[]]},{"id":"b9e5cd2c.db49b","type":"function","z":"5882ea6c.1d8c3c","name":"send markdown","func":"var message = msg.payload;\nmsg.payload = {chatId : 123456789, type : 'message', content : message};\n\n// activate markdown\nmsg.payload.options = {disable_web_page_preview : true, parse_mode : \"Markdown\"};\nreturn msg;","outputs":1,"noerr":0,"initialize":"","finalize":"","libs":[],"x":620,"y":80,"wires":[["13474f6d.cfac91"]]},{"id":"316ec0eb.a1532","type":"function","z":"5882ea6c.1d8c3c","name":"Alarm","func":"if(msg.payload.hookType !== undefined){\n    \n    if(msg.payload.hookType === 'motion_on' ){\n        msg.payload = \"Bewegunsgmelder Bewegung erkannt.\";\n        return msg;\n    }\n    \n    if(msg.payload.hookType === 'tamper_alarm_on' ){\n        msg.payload = \"Bewegunsgmelder Erschütterung erkannt.\";\n        return msg;\n    }\n}","outputs":1,"noerr":0,"initialize":"","finalize":"","libs":[],"x":350,"y":60,"wires":[["b9e5cd2c.db49b"]]},{"id":"99881d9e.bac2b","type":"function","z":"5882ea6c.1d8c3c","name":"Alarm","func":"if(msg.payload.hookType !== undefined){\n    \n    if(msg.payload.hookType === 'open_url' ){\n        msg.payload = \"Garagentuere geöffnet\";\n        return msg;\n    }\n    \n    if(msg.payload.hookType === 'close_url' ){\n        msg.payload = \"Garagentuere geschlossen\";\n        return msg;\n    }\n}\n","outputs":1,"noerr":0,"initialize":"","finalize":"","libs":[],"x":350,"y":120,"wires":[["b9e5cd2c.db49b"]]},{"id":"d4418730.706ad8","type":"shelly-gen1","z":"5882ea6c.1d8c3c","hostname":"192.168.178.250","description":"Motion","mode":"callback","server":"6511226786e4dc94","outputmode":"event","uploadretryinterval":"5000","pollinginterval":5000,"pollstatus":false,"getstatusoncommand":true,"devicetype":"Sensor","outputs":1,"credentials":{},"x":110,"y":60,"wires":[["316ec0eb.a1532","b2d978c31b0e54c7"]]},{"id":"1a21f56.8cc530b","type":"shelly-gen1","z":"5882ea6c.1d8c3c","hostname":"192.168.178.249","description":"Door","mode":"callback","server":"6511226786e4dc94","outputmode":"event","uploadretryinterval":"5000","pollinginterval":"5000","pollstatus":false,"getstatusoncommand":true,"devicetype":"Sensor","outputs":1,"credentials":{},"x":110,"y":120,"wires":[["99881d9e.bac2b","b2d978c31b0e54c7"]]},{"id":"746f7e11e7babe85","type":"shelly-gen1","z":"5882ea6c.1d8c3c","hostname":"192.168.178.120","description":"Heizung","mode":"callback","server":"6511226786e4dc94","outputmode":"event","uploadretryinterval":"5000","pollinginterval":"5000","pollstatus":false,"getstatusoncommand":true,"devicetype":"Measure","outputs":2,"credentials":{},"x":120,"y":180,"wires":[["a7db6c626a3fb297","b2d978c31b0e54c7"],[]]},{"id":"a7db6c626a3fb297","type":"function","z":"5882ea6c.1d8c3c","name":"Alarm","func":"if(msg.payload.hookType !== undefined){\n    \n    if(msg.payload.hookType === 'over_power_url' ){\n        msg.payload = \"Heizung: anheizen\";\n        return msg;\n    }\n    \n    // if(msg.payload.hookType === 'under_power_url' ){\n    //     msg.payload = \"Heizung: anheizen beendet\";\n    //     return msg;\n    // }\n}\n","outputs":1,"noerr":0,"initialize":"","finalize":"","libs":[],"x":350,"y":160,"wires":[["b9e5cd2c.db49b"]]},{"id":"59e305c4.e5a97c","type":"telegram bot","botname":"MyBot","usernames":"","chatids":"","baseapiurl":"","updatemode":"polling","pollinterval":"300","usesocks":false,"sockshost":"","socksport":"6667","socksusername":"anonymous","sockspassword":"","bothost":"","botpath":"","localbotport":"8443","publicbotport":"8443","privatekey":"","certificate":"","useselfsignedcertificate":false,"sslterminated":false,"verboselogging":false},{"id":"6511226786e4dc94","type":"shelly-gen1-server","port":"10002","hostname":"192.168.178.200"}]

**Reply by Fr3d3rick83** [reply]
Kann ich leider nicht importieren, da muss irgend wo ein Fehler in der Syntax sein.

**Comment by windkh** [comment, marked as answer]
[flow.txt](https://github.com/windkh/node-red-contrib-shelly/files/10167440/flow.txt)

**Reply by Fr3d3rick83** [reply]
Funktioniert, jetzt muss ich es mir nur noch al komplette Alarmanlage zusammen bauen. TAUSEND DANK!
